lib/puppet/transport/cisco_ios.rb

```diff
@@ -0,0 +1,370 @@
+require_relative '../../puppet_x/puppetlabs/cisco_ios/utility'
+
+module Puppet::Transport
+  # Provides a Transport for making remote calls to a Cisco Ios device
+  class CiscoIos
+    # configuration state, eg in tacacs mode
+    class ModeState
+      NOT_CONNECTED = 1 unless defined? NOT_CONNECTED
+      LOGGED_IN = 2 unless defined? LOGGED_IN
+      ENABLED = 3 unless defined? ENABLED
+      CONF_T = 4 unless defined? CONF_T
+      CONF_INTERFACE = 5 unless defined? CONF_INTERFACE
+      CONF_TACACS = 6 unless defined? CONF_TACACS
+      CONF_VLAN = 7 unless defined? CONF_VLAN
+      CONF_TACACS_SERVER_GROUP = 8 unless defined? CONF_TACACS_SERVER_GROUP
+      CONF_RADIUS_SERVER_GROUP = 9 unless defined? CONF_RADIUS_SERVER_GROUP
+      CONF_RADIUS_SERVER = 10 unless defined? CONF_RADIUS_SERVER
+      CONF_LINE = 11 unless defined? CONF_LINE
+      CONF_MST = 12 unless defined? CONF_MST
+      CONF_STD_NACL = 13 unless defined? CONF_STD_NACL
+      CONF_EXT_NACL = 14 unless defined? CONF_EXT_NACL
+    end
+
+    attr_reader :connection, :config
+    attr_accessor :commands
+
+    def initialize(context, config, options = {})
+      @context = context
+      @config = config
+
+      create_connection(options[:debug])
+      @enable_password = config[:enable_password].unwrap
+      @facts = facts(context)
+      PuppetX::CiscoIOS::Utility.facts(@facts)
```

Comment:
This seems backwards. Any facts processing should go to the `facts` method

Reply:
This is copy pasta from device.rb. It appears that this initialises the `Utility` class with device facts that it needs for some of its methods to function, which is why it's performed during the initialisation of the transport.

Reply:
https://github.com/puppetlabs/cisco_ios/blob/42d93a6fffd8794dd4760d35af9dec2380a83133/lib/puppet_x/puppetlabs/cisco_ios/utility.rb#L46

ugh